### ISSUE TITLE
Jest fails with `Couldn't find temp query result` error

### ISSUE THREAD

**Issue by ascorbic** [gatsbyjs/gatsby#35018]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

As part of a site's unit tests, it may need to compile the site. Previously this was simple to do by running the cli via `child_process` or `execa`. This works until  `gatsby@4.9.0-next.0`, but since `gatsby@4.9.0-next.1` this fails. The site builds correctly if build manually from the CLI.

### Reproduction Link

https://github.com/ascorbic/gatsby-jest-bug

### Steps to Reproduce

1. Clone the repo
2. Run yarn install
3. Run yarn test

### Expected Result

Site builds and test passes

### Actual Result

Site build fails with the following error:

```
success Rewriting compilation hashes - 0.001s

     ERROR

    UNHANDLED REJECTION Couldn't find temp query result for "/404/".



      Error: Couldn't find temp query result for "/404/".

      - page-data.ts:104 readPageQueryResult
        [jest-demo]/[gatsby]/src/utils/page-data.ts:104:11

      - page-data.ts:121 writePageData
        [jest-demo]/[gatsby]/src/utils/page-data.ts:121:18

      - page-data.ts:232
        [jest-demo]/[gatsby]/src/utils/page-data.ts:232:24
```

### Environment

```shell
System:
    OS: macOS 11.6.4
    CPU: (8) arm64 Apple M1
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 16.11.0 - /var/folders/k2/ffw3_41j7h17rzsmx25wcwl80000gn/T/yarn--1646216980464-0.9843919045068601/node
    Yarn: 1.22.17 - /var/folders/k2/ffw3_41j7h17rzsmx25wcwl80000gn/T/yarn--1646216980464-0.9843919045068601/yarn
    npm: 8.3.0 - ~/.nvm/versions/node/v16.11.0/bin/npm
  Languages:
    Python: 2.7.16 - /usr/bin/python
  Browsers:
    Chrome: 98.0.4758.102
    Safari: 14.1.2
  npmPackages:
    gatsby: ^4.9.0 => 4.9.0
```

**Comment by JeffersonCarvalh0** [comment]
I'm also experiencing this problem with gatsby@4.8.0

**Comment by Nowaker** [comment]
Same for me.

```
es-createPages: 39.095s
success createPages - 89.772s
success createPagesStatefully - 0.051s
info Total nodes: 6030, SitePage nodes: 1642 (use --verbose for breakdown)
verbose 99.876148055 Number of node types: 10. Nodes per type: Directory: 750, File: 1801, JekyllConfigBundled: 1, JekyllData: 647, JekyllDataBundled: 1, JekyllHtml: 1154, Site: 1, SiteBuildMetadata: 1, SitePage: 1642, SitePlugin: 32
verbose 99.876242955 Checking for deleted pages
verbose 99.877802396 Deleted 0 pages
verbose 99.878584229 Found 1642 changed pages
success Checking for changed pages - 0.001s
success onPreExtractQueries - 0.000s
success extract queries from components - 1.687s
success write out redirect data - 0.004s
success onPostBootstrap - 0.001s
info bootstrap finished - 101.578s
success write out requires - 0.015s
success Building production JavaScript and CSS bundles - 5.234s
success Building HTML renderer - 9.245s
success Execute page configs - 0.019s
success Caching Webpack compilations - 0.001s
success run queries in workers - 2.095s - 1642/1642 783.85/s
success Merge worker state - 0.007s
success Rewriting compilation hashes - 0.001s
error UNHANDLED REJECTION Couldn't find temp query result for "/es/thank-you/index.html".
  Error: Couldn't find temp query result for "/es/thank-you/index.html".
  
  - page-data.ts:104 readPageQueryResult
    [_gatsby]/[gatsby]/src/utils/page-data.ts:104:11
  
  - runMicrotasks
  
  - task_queues:61 runNextTicks
    node:internal/process/task_queues:61:5
  
  - timers:437 processImmediate
    node:internal/timers:437:9
  
  - page-data.ts:121 writePageData
    [_gatsby]/[gatsby]/src/utils/page-data.ts:121:18
  
  - page-data.ts:232 
    [_gatsby]/[gatsby]/src/utils/page-data.ts:232:24
  
not finished Writing page-data.json files to public directory - 4.315s
```

Env: Linux (CI), Gatsby 4.7.2.

Note: it started happening recently. I haven't seen this error before. No changes were done to the site in between. CI just started encountering this problem out of the blue moon. Trying `yarn install --frozen-lockfile` now to see if that fixes the problem. In which case a diff between by locked dep resolution and most recent dep resolution will be the issue. Otherwise... who knows.

**Comment by Nowaker** [comment]
`yarn install --frozen-lockfile` helped.

**Comment by marciolima-praycom** [comment]
I'm experiencing the same issue with `gatsby: ^4.11.0`. 

Running `yarn install --frozen-lockfile` doesn't help.. 😔

**Comment by ZeldOcarina** [comment]
Same issue here! :(
I am 4.14.0

**Comment by Oakwhisper** [comment]
Creating a new blank gatsby project builds locally on MacOS Monterrey (ARM & x64) but fails on jenkins server build (Debian 11).
Error is:
```
success Merge worker state - 0.003s
success Rewriting compilation hashes - 0.062s
error UNHANDLED REJECTION Couldn't find temp query result for "/404/".

[0m
[0m  [0m[97m[41mError[0m[37m[41m:[0m[37m[41m [0m[97m[41mCouldn't find temp query result for "/404/".[0m
[0m  [0m
[0m  [0m[90m-[0m [0m[93mpage-data.ts[0m[90m:[0m[93m104[0m[37m [0m[37mreadPageQueryResult[0m
[0m  [0m  [0m[90m[gatsby-test-project]/[gatsby]/src/utils/page-data.ts:104:11[0m
[0m  [0m
[0m  [0m[90m-[0m [0m[93mpage-data.ts[0m[90m:[0m[93m121[0m[37m [0m[37mwritePageData[0m
[0m  [0m  [0m[90m[gatsby-test-project]/[gatsby]/src/utils/page-data.ts:121:18[0m
[0m  [0m
[0m  [0m[90m-[0m [0m[93mpage-data.ts[0m[90m:[0m[93m232[0m[37m [0m
[0m  [0m  [0m[90m[gatsby-test-project]/[gatsby]/src/utils/page-data.ts:232:24[0m
[0m  [0m
[0m
not finished Writing page-data.json files to public directory - 0.408s
```

Dependencies for brand new gatsby project: 
```
  "dependencies": {
    "@emotion/react": "^11.9.0",
    "@emotion/styled": "^11.8.1",
    "gatsby": "^4.14.1",
    "gatsby-plugin-emotion": "^7.14.0",
    "gatsby-plugin-manifest": "^4.14.0",
    "react": "^17.0.1",
    "react-dom": "^17.0.1"
  }
```

**Comment by enriqueArb** [comment]
I had this issue too with v4.12.1.
For me, the solution was to run `gastby clean`. After that, build ran correctly.

Hope it helps someone.

**Comment by belfz** [comment]
Same on gatsby v4.15.2. Neither `gatsby clean` nor `yarn install --frozen-lockfile` helped. The error also feels very random, each time mentioning other page within the `Couldn't find temp query result for (...)` error message.

**Comment by belfz** [comment]
@LekoArts here's a URL to the faulty project (minimal reproduction): https://github.com/belfz/gatsby-bug-reproduce

**Comment by LekoArts** [comment]
@belfz Your reproduction shows the problem because you're using the `createPage` action in the wrong lifecycle hook. It's only allowed in `createPages` while you use it in `onCreateWebpackConfig`.

@ascorbic The PR https://github.com/gatsbyjs/gatsby/pull/35968 should fix this, wasn't easy to spot this 😅

**Comment by ascorbic** [comment]
Ooh, good find! Sneaky one

**Comment by Nowaker** [comment]
I wouldn't consider it resolved. My use case doesn't involve executing createPage in onCreateWebpackConfig.

**Comment by belfz** [comment]
big thank you @LekoArts, indeed that was it - I didn't spot that I was using a wrong lifecycle hook 😅

**Comment by beamercola** [comment]
@Nowaker Agreed - I'm getting this intermittently on netlify deploys and I'm not doing anything funny with `createPage`.  Did you resolve your problem?

**Comment by Nowaker** [comment]
@beamercola Not really. Going back to some older lockfile did resolve most of it, but not all. It happens every once in a while.

**Comment by quanglam2807** [comment]
We're facing the same issue when running `build` on Netlify.

**Comment by zolon4** [comment]
I'm getting the same issue on netlify when running `build`. Sometimes clearing the netlify cache fixes but not always.

**Comment by enormousrodent** [comment]
I am also getting netlify deploy issues during build.  I have to keep re-building then it will eventually work!  really annoying and every time its a different page it complains about!

**Comment by jlevy-io** [comment]
Same here, this has been an issue on Netlify since August 1st 

(@netlify/build 27.8.1 is when we first started seeing it happen)

**Comment by natapg** [comment]
same here

**Comment by marciolima-praycom** [comment]
In my case I have a monorepo with 3 Gatsby apps and I was trying to build them on Circle CI using Lerna or Yarn Workspaces to call the clean & build scripts in all of them at once (either sequentially or in parallel). I was seeing that issue happening randomly for different pages.

I was able to fix it by creating a `build.sh` file that calls `yarn clean && yarn build` on each app separately. The Circle CI job now calls `yarn build` on my monorepo, and the build script calls the `build.sh` file like this:

```json
"scripts": {
  "build": "yarn exec ./build.sh"
},
```

Not sure why, but after we moved to this approach we never saw that error happening again.

**Comment by natapg** [comment]
I run `yarn clean && yarn build `on every run due to another problem, and I still started having this issue a few days ago

**Comment by mitchossi** [comment]
Same problem here,` gatsby clean` didn't help...

**Comment by NeversSync** [comment]
Having the same issue as well, very intermittent and frustrating

**Comment by coreyward** [comment]
Another +1 here. Sometimes also seeing PQR related errors—could be related, or a separate issue.

**Comment by ZeldOcarina** [comment]
> Another +1 here. Sometimes also seeing PQR related errors—could be related, or a separate issue.

Hey what is "pqr"?

**Comment by coreyward** [comment]
+1000000

@LekoArts Is calling `createPage` in `onCreatePage` still okay? I'm using it like the [docs show here](https://www.gatsbyjs.com/docs/creating-and-modifying-pages/#pass-context-to-pages).

**Comment by coreyward** [comment]
I just ran into this in `gatsby develop` mode too. Using 4.21.0.

**Comment by Nowaker** [comment]
It had nothing to do for me with Netlify. Happening to me in build. To others in develop. This is a random bug in Gatsby - and it happens so ramdomly that there's no reproduction, let alone a minimal one.

**Comment by jasonlav** [comment]
I experienced this issue when building pages using nested methods. Removing the nesting appears to have resolved the issue.

**Example of what caused the issue** 
```
export.createPages = async (args) => {
    await createPageTypeA(args);
};

async function createPageTypeA(args) {
    await createPageInUniversalSharedMethod(args);
}

async function createPageInUniversalSharedMethod({ createPage }) {
   ....
   createPage(...);
    ...
}
```

**Example of what resolved the issue** 
```
export.createPages = async (args) => {
    await createPageInUniversalSharedMethod(args);
};

async function createPageInUniversalSharedMethod({ createPage }) {
   ...
   createPage(...);
   ...
}
```

**Comment by coreyward** [comment]
@jasonlav Interesting. Any chance it's a fluke that your issue is fixed?

@LekoArts @graysonhicks Is Gatsby doing any kind of static analysis on gatsby-node that might cause this?

**Comment by jasonlav** [comment]
> Interesting. Any chance it's a fluke that your issue is fixed?

Absolutely could be. However, I have not encountered the issue since that change.

**Comment by cesarvarela** [comment]
it has nothing to do with Jest IMO, we started getting this error after upgrading gatsby

**Comment by joebentaylor1995** [comment]
Same issue, not using jest or even creating pages yet

**Comment by aaadamgo** [comment]
This is still a problem on @4.24.1

I've remove node modules, gatsby clean, etc and try to build locally and get the error `UNHANDLED REJECTION Couldn't find temp query result for "/404/".`

I was having issues on @4.12.1 but then upgraded to the above to see if a fix was in place.

My 404 page has nothing on it.

Tried removing the 404, it then pointed to another page, same error, different page.

Sorry this is a bit of a 'live' update as I'm working through it.

I've stepped backwards through the gatsby versions, I went to 4.8.1 and the error still happens. I just tried jumping to @4.2, and it gave me a better error (turned out to be a graphql query on a completely separate page). I fixed that and now it builds perfectly fine locally.

Now circling back to @4.24.1

Error'd again. So it's between 4.2 and 4.8.

@4.6 seems to get stuck on the building pages process. I gave it 10 minutes (I only have 4 pages).

Tried all versions between @4.2 and @4.6, and only @4.2 will build the pages.

**Comment by graysonhicks** [comment]
Could someone please post a minimal reproduction that can be run locally?

**Comment by ZeldOcarina** [comment]
Unfortunately this minimal reproduction is a huge stop on handling major bugs. It's often really not possible.
What about entering the existing repo? I think that is soo much easier on both parts (when possible of course)

**Comment by coreyward** [comment]
@graysonhicks I can't reproduce this consistently with our full app and we've been suffering from it for weeks. If the error message was able to narrow down the source it might be viable to provide a minimal reproduction, but as it stands, I don't think it's plausible outside of cases where it arrives through obvious user error.

For the record, I tried a solution similar to the one posted by @jasonlav above and I am still running into this.

**Comment by Nowaker** [comment]
Minimal reproduction example is a pipe dream when a problem is related to concurrency and manifests itself only in non-toy projects.

**Comment by ZeldOcarina** [comment]
Yeah and just so you know, this exact problem drove me to using Jenkins and dedicated servers and even local builds instead of Gatsby Cloud as it was always failing there.. and working on my local mac..

**Comment by cesarvarela** [comment]
Here is the error in all its glory; it happens at "random" and usually is fixed by rerunning the build with an empty cache:

https://app.netlify.com/sites/aiid/deploys/633bd0132ef6160009949592

(I know, I know, netlify, but this is the example I have at hand)

The repo: https://github.com/responsible-ai-collaborative/aiid

My spider sense tells me that it has to do with concurrency/size, and a minimal repro case is impossible.

**Comment by graysonhicks** [comment]
@coreyward @ZeldOcarina @cesarvarela Thanks, I know it's frustrating, it's just near impossible for us to address without a reproduction locally.  Do any of you have links to a build where this is happening on Gatsby Cloud?  I have more tools available on that end, but I don't see any links to builds on Gatsby Cloud, or any links to reproductions that happen locally.

**Comment by ZeldOcarina** [comment]
I do!
https://www.gatsbyjs.com/dashboard/e968afff-0cff-4103-8d02-5dbbe05c6a6f/sites/26587904-b4fa-402f-b570-2cb1bbabca80/builds/61d86c44-1010-4154-a00e-f82d681b9823/details#rawLogs
If you check inside that project there are so many of these builds failing while locally these are working!

**Comment by graysonhicks** [comment]
Thanks @ZeldOcarina!  I'll open up a ticket on the Gatsby Cloud side and take a look.  Do you mind sending an email to support@gatsbyjs.com and tag me in that email and we can track it that way and report back on this thread.

**Comment by graysonhicks** [comment]
Actually @ZeldOcarina I don't see the `Couldn't find temp query result` error in any of those builds...

**Comment by graysonhicks** [comment]
@ZeldOcarina For the builds you sent, they are OOM'ing.  Often, it OOMs and exits before the error can be written to the logs.  Without looking at the code for a while, I can't say why exactly (at a glance I think adding `width` option to your image queries would help), but [this doc is useful for improving memory consumption](https://www.gatsbyjs.com/docs/how-to/performance/resolving-out-of-memory-issues/).

![image](https://user-images.githubusercontent.com/16200795/194155598-20a9a100-2638-403f-9687-92c59ac539d4.png)

But I still don't see an instance of the `temp query result` error and I don't think memory usage is related.  When I bumped that site to 8GB of memory (from 4GB) it builds successfully.

**Comment by ZeldOcarina** [comment]
Hi @graysonhicks,
You are right it is failing out of memory heap as it has so many GatsbyImages.
Actually I couldn't handle that error so I just used normal img tags unfortunately on all next projects.
When building locally it sometimes builds and it sometimes fails with cannot find property "id" of null, with appearently no resons.
Weird..

**Comment by TRegourd** [comment]
I am also experiencing it, randomly. It's a shame because my site is supposed to rebuild automatically on a certain interval, but one time in two I have to clear cache and relaunch the build myself ...

**Comment by cesarvarela** [comment]
I've had some success setting `GATSBY_CPU_COUNT=2`.

I guess that it limits how much stuff Gatsby tries to do at once, limiting memory consumption.

Also, if you are on CI and have a limited number of cores, it leaves free cores for other processes, like Cypress, local servers, etc.

**Comment by hotfix519** [comment]
> I've had some success setting `GATSBY_CPU_COUNT=2`.

I don't want to jinx it, b/c it's still not a terribly large sample size, but this has shown improvements for us.  I was getting the temp query failure intermittently, but it's been clear for a bit. 🤞 It doesn't seem to significantly affect build times either, which was an initial concern.

I'll report back if it pops back up or stays clear over a larger sample size.

**Comment by Nowaker** [comment]
Limiting the concurrency is a very bad workaround. The whole selling point of Node.js and Gatsby is the fact it's highly concurrent. No or low concurrency is like using Jekyll - a huge downgrade.

**Comment by cesarvarela** [comment]
It is a bad workaround depending on how much it affects your build time, which in my case, was not that much different.

**Comment by hotfix519** [comment]
Also, I don't think any of us are suggesting that this issue can be closed or that this "solves" it.  But perhaps it can allow some of us that are stuck w/ these errors to move forward in the interim.

**Comment by jfgilmore** [comment]
> Also, I don't think any of us are suggesting that this issue can be closed or that this "solves" it. But perhaps it can allow some of us that are stuck w/ these errors to move forward in the interim.

This issue has been closed though. @cesarvarela would you be able to put your repo and reproduction details in a new issue? It doesn't look like anyone has done this as yet for builds, this issue was related to Jest. Although I don't think Jest had anything to do with it I believe a new issue needs to be opened and linked here as people are still experiencing this problem.

Unfortunately the repo I am having the issue with is private 😞

**Comment by hotfix519** [comment]
> Unfortunately the repo I am having the issue with is private disappointed

Same here.  I do feel like it has something to do with larger amts of data/pages.  I wonder if we tried to make a really basic project that's pushing a TON of faked data thru the gatsby pipeline and creating thousands of pages, maybe we can get sth public.  I may give it a shot if I have time, sad thing is, I rarely do :disappointed:

**Comment by coreyward** [comment]
As an additional data point, I too am observing relief with `GATSBY_CPU_COUNT=4` and no discernible difference in build times on Netlify. This makes me wonder if this is partly related to resource contention of virtualized CPUs.

**Comment by Nowaker** [comment]
> This makes me wonder if this is partly related to resource contention of virtualized CPUs.

Even if a problem manifests itself when resource use is high, the core issue in such situations is race conditions caused by buggy concurrent code. It's easy to make bugs even in green threads environment when working with multiple variables, let alone when real workers are involved.

**Comment by ZeldOcarina** [comment]
Hey @Nowaker what is a "green thread" environment?

**Comment by Nowaker** [comment]
@ZeldOcarina Green threads: https://en.wikipedia.org/wiki/Green_thread - and Node.js itself uses green threads for "concurrency", while some other language may or may not use it. "Concurrency" and not concurrency because Node.js has GIL: https://en.wikipedia.org/wiki/Global_interpreter_lock, though it's still more concurrent than other languages with GIL because I/O waits put its green thread in sleep mode, and a different non-blocked green thread is running.

**Comment by ZeldOcarina** [comment]
@Nowaker wow what a deep insight, thank you so much!

**Comment by iamabdullahmalik** [comment]
Did anyone find a solution for this?
I have the same issue with 5.0, I have almost 60k posts and 20GB of images.

**Comment by joshuar500** [comment]
for netlify, creating an environment variable `GATSBY_CPU_COUNT=2` worked for me

**Comment by jfgilmore** [comment]
> for netlify, creating an environment variable `GATSBY_CPU_COUNT=2` worked for me

I have this in place but am still getting the error, just less frequently than before (we used to have around 50% of builds fail, now it's probably more like 5%).

**Comment by flangham** [comment]
Am also having this issue.

**Comment by dougkellermeyer** [comment]
Still observing this issue - random pages error as @belfz mentioned. Updated my build Netlify commands to `gatsby clean && gatsby build` and issue seems to have resolved itself. 

I built a custom front end used the Gatsby Shopify Starter - https://github.com/gatsbyjs/gatsby-starter-shopify 

Also, feel like it's worth mentioning that I experience the same error for local builds and running `gatsby clean` solves it as well 🤷

**Comment by techbecca** [comment]
I am also experiencing this issue, `gatsby clean` solved it.

**Comment by julianCast** [comment]
Just a reminder, this is still happening.  We can't provide a reproducible sample since it appears randomly (racing condition? memory issue?)
```log
1:21:37 PM: error Failed to write page-data for ""/it-it/blog/page/ Couldn't find temp query result for "/it-it/blog/page/".
1:21:37 PM: Error: Couldn't find temp query result for "/it-it/blog/page/".
```

**Comment by julianCast** [comment]
I dont think adding `gatsby clean` to the build command is  a proper solution, it will delay each build.
